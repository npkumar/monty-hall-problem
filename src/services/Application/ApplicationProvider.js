import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";

// Actions
const SET_SIMULATIONS = "SET_SIMULATIONS";
const SET_IS_CHANGE_DOOR = "SET_IS_CHANGE_DOOR";
const SET_WIN_PERCENTAGE = "SET_WIN_PERCENTAGE";
const SET_IS_SIMULATION_RUNNING = "SET_IS_SIMULATION_RUNNING";

export const ApplicationContext = createContext(null);

function reducer(state, action) {
  switch (action.type) {
    case SET_SIMULATIONS: {
      return { ...state, simulations: action.payload, winPercentage: null };
    }
    case SET_IS_CHANGE_DOOR: {
      return { ...state, isChangeDoor: action.payload, winPercentage: null };
    }
    case SET_WIN_PERCENTAGE: {
      return { ...state, winPercentage: action.payload };
    }
    case SET_IS_SIMULATION_RUNNING: {
      return { ...state, isSimulationRunning: action.payload };
    }
    default:
      throw new Error();
  }
}
const ApplicationProvider = ({ children }) => {
  const initialState = {
    simulations: 4000,
    isChangeDoor: true,
    winPercentage: null,
    isSimulationRunning: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const value = {
    application: state,
    setSimulations: (simulations) =>
      dispatch({ type: SET_SIMULATIONS, payload: simulations }),
    setIsChangeDoor: (isChangeDoor) =>
      dispatch({ type: SET_IS_CHANGE_DOOR, payload: isChangeDoor }),
    setWinPercentage: (winPercentage) =>
      dispatch({ type: SET_WIN_PERCENTAGE, payload: winPercentage }),
    setIsSimulationRunning: (isSimulationRunning) =>
      dispatch({
        type: SET_IS_SIMULATION_RUNNING,
        payload: isSimulationRunning,
      }),
  };

  return (
    <ApplicationContext.Provider value={value}>
      {children}
    </ApplicationContext.Provider>
  );
};

ApplicationProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ApplicationProvider;
