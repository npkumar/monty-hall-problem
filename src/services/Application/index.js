// Hooks
export { default as useApplication } from "./useApplication";

// Providers
export { default as ApplicationProvider } from "./ApplicationProvider";
