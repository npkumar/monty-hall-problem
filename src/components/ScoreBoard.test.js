import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import Scoreboard from "./Scoreboard";
import { ApplicationContext } from "../services/Application/ApplicationProvider";

const value = {
  application: { isSimulationRunning: false, winPercentage: 10 },
};

describe("Scoreboard", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const { asFragment } = render(
      <ApplicationContext.Provider value={value}>
        <Scoreboard />
      </ApplicationContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should a scoreboard percentage", () => {
    const { getByTestId } = render(
      <ApplicationContext.Provider value={value}>
        <Scoreboard />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("percentage")).toBeInTheDocument();
    expect(getByTestId("percentage")).toHaveTextContent("10%");
  });
});
