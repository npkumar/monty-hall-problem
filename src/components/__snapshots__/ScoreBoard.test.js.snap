// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard renders 1`] = `
<DocumentFragment>
  <div
    class="MuiTypography-root makeStyles-root-1 MuiTypography-body1"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5"
        >
          Win Percentage
        </h5>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
      style="margin-left: 20px;"
    >
      <div
        class="MuiCardContent-root"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4"
          data-testid="percentage"
        >
          10%
        </h4>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
